Court: Polish main court contract functionality

facuspagnuolo opened the pull request:
Fixes #83 and #16 

Lots of polishing stuff, inline documentation and bunch of new tests happening on this PR, important pieces to take a look carefully are:

- ⚠️ Ensure first term is in the future (https://github.com/aragon/aragon-court/pull/103/commits/53ff9e31fa54cc4ef513342d6c2f29e6003b5e02#diff-4bb317b489994cf6f4c4b7ed11c18590R235)
- ⚠️ Ensure term is not zero when creating disputes (https://github.com/aragon/aragon-court/pull/103/commits/79abd2686bfdada8fbd69badb7453a08db0bee38#diff-4bb317b489994cf6f4c4b7ed11c18590R297)
- ⚠️ Ensure appealer cannot confirm an appeal (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R456)
- ⚠️ Use start term id of the final round to fetch the total active balance of the jurors' registry when computing the jurors number (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R1294)
- Unify appeal struct (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R172-R177)
- Handle `Pre-Draft` and `Executed` dispute states when querying an adjudication round state (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R1362-R1371)
- Add round getter (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R760)
- Add appeal getter (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R796)
- Update `settledJurors` even for the last batch to ensure consistency (https://github.com/aragon/aragon-court/pull/103/commits/2e3cbbd45b1277ba6d27f7ca30e021b1dc85713a#diff-4bb317b489994cf6f4c4b7ed11c18590R1008)
- New thousands percentage lib (https://github.com/aragon/aragon-court/pull/103/commits/0cb4d62a7597819478a7bc5723b1611fdb074eb6?file-filters%5B%5D=.sol#diff-7c2369ae5667d560b954f85000fff662R6-R22)

izqui commented on contracts/Court.sol:
```diff
@@ -233,8 +233,7 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         require(_firstTermStartTime >= _termDuration, ERROR_BAD_FIRST_TERM_START_TIME);
-        // TODO: we should add this validation, cannot enable it now due to how tests are mocking timestamps
-        // require(_firstTermStartTime - _termDuration <= getTimestamp64(), ERROR_BAD_FIRST_TERM_START_TIME);
+        require(_firstTermStartTime >= getTimestamp64(), ERROR_BAD_FIRST_TERM_START_TIME);
```
How about collapsing it into just one check: `_firstTermStartTime >= getTimestamp64() + _termDuration`?

facuspagnuolo replied:
Sure! I like the idea

facuspagnuolo replied:
Fixed in daffddea0c5a2162f2dce1f673edd5e685469a35

izqui commented on contracts/Court.sol:
```diff
@@ -652,38 +670,34 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+            Term storage previousTerm = terms[termId++];
+            Term storage nextTerm = terms[termId];
+
+            // TODO: allow config to be changed for a future term id
```
🚨 This is a pretty important TODO, as right now it is impossible for the configuration to change from term to term.

facuspagnuolo replied:
Yup, I was planning to tackle all together in #53

facuspagnuolo commented:
@izqui GitHub is not showing any code related in your comments, would you mind specifying a link for them? At least for the first and the third one, I can imagine what you referred to with the second one :)

izqui commented:
@facuspagnuolo I was committing on the commits directly by mistake, will add the comments to the review directly.

facuspagnuolo commented on package.json:
```diff
@@ -4,17 +4,16 @@
-    "analysis:tree": "TRUFFLE_TEST=true SUMTREE_GAS_ANALYSIS=true npm run ganache-cli:test",
     "coverage": "SOLIDITY_COVERAGE=true npm run ganache-cli:test",
-    "ganache-cli:test": "./node_modules/@aragon/test-helpers/ganache-cli.sh",
+    "ganache-cli:test": "./scripts/ganache-cli.sh",
```
Now that we released a new minor yes, but it had a bug that didn't allow me to split the CI into multiple tasks

facuspagnuolo replied:
Fixed in 4f4e205bd34c992dad8e9867d9e5d5116c01875d

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +335,154 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
     function appeal(uint256 _disputeId, uint256 _roundId, uint8 _ruling) external ensureTerm {
-        _checkAdjudicationState(_disputeId, _roundId, AdjudicationState.Appeal);
-
+        // TODO: ensure round exists
```
I've been having some size issues so I stopped adding these to some functions, we can check after merging the solc upgrade how it goes

facuspagnuolo replied:
I think so, and some stack too deep issues

bingen replied:
https://github.com/aragon/aragon-court/pull/103#discussion_r321189927
:thinking: :stuck_out_tongue_winking_eye:

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +335,154 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-            jurorsRequested = MAX_JURORS_PER_DRAFT_BATCH;
-        }
+        // Draft the min number of jurors between the one requested by the sender and the one requested by the disputer
+        uint64 jurorsNumber = round.jurorsNumber;
```
Good catch! I added a test for this and it reverts because it ends trying a sortition out of bounds which makes sense

facuspagnuolo replied:
Fixed in 4adbc9b26321c43610110f022537d17d5e68d742

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +335,154 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         // TODO: ERC165 check that _subject conforms to the Arbitrable interface
-
         // TODO: require(address(_subject) == msg.sender, ERROR_INVALID_DISPUTE_CREATOR);
+        require(termId > ZERO_TERM_ID, ERROR_CANNOT_CREATE_DISPUTE);
```
Sure, I didn't want to add that yet cause I think there was an ongoing discussion to define a minimum amount of terms in the future for each dispute to ensure there is enough time to provide evidence.

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1139,415 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        )
+    {
+        // Next round start term is current round end term, nevertheless it is final or regular
+        nextRoundStartTerm = _getRegularRoundEndTerm(_round);
```
True! Fixed in c0ed4d61a1dcccf07ea74c358b87ad1701f2edde

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -780,349 +657,431 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @return Number of terms the Court should transition to be up-to-date
+    */
+    function neededTermTransitions() public view returns (uint64) {
+        // Note that the Court is always initialized at least for the current initialization time or more likely a
```
True, fixed in aadd9735ef0100145eaa93fbbeb55f62c468caf0

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +335,154 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        // Ensure given round appeal was not confirmed yet
+        AdjudicationRound storage round = dispute.rounds[_roundId];
         require(_isRoundAppealed(round), ERROR_ROUND_NOT_APPEALED);
+        require(round.appeal.maker != msg.sender, ERROR_SENDER_CANNOT_CONFIRM_APPEAL);
```
Yea, it doesn't make sense. Fixed in da166b3b5d5fc1f29c999c574a5f8218573f022f

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -780,349 +657,431 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @return weight Juror weight drafted for the requested round
+    * @return rewarded Whether or not the given juror was rewarded based on the requested round
+    */
+    function getJuror(uint256 _disputeId, uint256 _roundId, address _juror) external roundExists(_disputeId, _roundId) view
```
Definitely it looks weird, didn't find any specification on this in the official style guide, let's put those modifiers together

bingen commented on contracts/Court.sol:
```diff
@@ -263,264 +334,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        // Draft jurors for the given dispute and reimburse fees
+        CourtConfig storage config = _getConfigAtDraftTerm(round);
+        _draft(jurorsRequested, _disputeId, round, draftTerm, config);
+        accounting.assign(config.feeToken, msg.sender, config.draftFee * jurorsRequested);
```
I see you have added SafeMath in some places but not here. Is this intended? Are we sure this doesn't overflow?

facuspagnuolo replied:
No, I only added it where I was sure we should use it, added comments in other places where we are safe to avoid it, and didn't change those that I wasn't sure yet. As we discussed offline  I think we should track all the math happening in the court and analyze and test each of them carefully. I was thinking we could do so in #47

bingen commented on contracts/Court.sol:
```diff
@@ -263,264 +334,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+
+        // If the drafting is over, update its state
+        if (round.selectedJurors == round.jurorsNumber) {
+            // Note that we can avoid using SafeMath here since we already ensured `termId` is greater than or equal to `round.draftTermId`
```
:+1:

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+        uint64 revealTerms;         // revealing period duration in terms
+        uint64 appealTerms;         // appealing period duration in terms
+        uint64 appealConfirmTerms;  // confirmation appeal period duration in terms
+        uint16 penaltyPct;          // percentage that will be used to compute the tokens to be locked for a juror on a draft
```
sure, let's use "permille"

facuspagnuolo replied:
Fixed in 5dccdf737b624d78cfcfc45ca9cc73ff19204f68

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1137,405 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-        require(_subscriptionParams[4] <= MAX_UINT16, ERROR_OVERFLOW); // _governorSharePct
+    // TODO: move to a factory contract
+    function _initSubscriptions(ERC20 _feeToken, uint256[5] _subscriptionParams) private {
+        uint64 MAX_UINT64 = uint64(-1);
```
I noted the initialization code size is considerable, about 5k... I created https://github.com/aragon/aragon-court/issues/78 to explore the idea of having sth like a factory contract. This is a simple step towards that direction while making sure those variables won't be held in the contract body

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +334,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-            }
+        // Draft jurors for the given dispute and reimburse fees
+        CourtConfig storage config = _getConfigAtDraftTerm(round);
+        _draft(jurorsRequested, _disputeId, round, draftTerm, config);
```
😂 sorry for that, yea it sounds ridiculous, let me see if I can come up with sth else

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+    * @dev Ensure a dispute exists
+    * @param _id Identification number of the dispute to be ensured
+    */
+    modifier disputeExists(uint256 _id) {
```
yeaaaa... we can argue whether to use modifiers or not, but I wouldn't use the size limit as an argument unless it becomes an issue, for now we are under the limit

bingen replied:
Well, that's the point, it's showing me now more than 25k of bytecode size (maybe I'm doing something wrong?). I know tests are passing, but we should try testing with a real node to be sure.
Besides, given the amount of headaches EIP-170 has given us during last months, how close we have been since then to the limit (if not above) and that this contract is still under heavy development, i think we should always keep it in mind.

facuspagnuolo replied:
Let's talk about this topic and agree on sth, I feel like we are spreading different opinions here and there about the same topic

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+    // Configs-related error messages
+    string private constant ERROR_INVALID_PENALTY_PCT = "CT_INVALID_PENALTY_PCT";
+    string private constant ERROR_INVALID_SENDER_ADDRESS = "CT_INVALID_SENDER_ADDRESS";
+    string private constant ERROR_BAD_FIRST_TERM_START_TIME = "CT_BAD_FIRST_TERM_START_TIME";
```
We are not above the limit, tests won't be passing if that was the case. IMO we should optimize only if necessary, do you agree?

bingen replied:
Nope, sorry.
As I said, I'm afraid we are quite over the limit.
Besides, I would say we are here "undoing an already existing optimization", which is slightly different. An optimization that was done at a moment that was absolutely necessary.
The only benefit for removing it being slightly longer messages that are not end-user intended, so even in the case we were not yet over the limit, I don't think the change is worth it.

bingen replied:
Update: I was using the wrong json key in `build/contracts/Court.json`, so you are right, we are under the limit. But still I don't see the point of this unnecessary strings enlargement.

facuspagnuolo replied:
Are we talking about a size optimization or a tx-cost one?

bingen replied:
I was talking about bytecode size. I haven't checked tx cost.

facuspagnuolo replied:
As mentioned above, let's discuss size optimizations in general and then apply whatever changes are needed

facuspagnuolo replied:
I'm not applying these changes, I'm proposing them and in fact, wasn't thinking of merging this PR until we have that conversation

bingen replied:
Ok, but I wouldn't block such a huge PR like this one for this (besides, we have several minor PRs behind waiting for this). I would just revert those changes, move forward with this PR, have the discussion and the open a separate PR for increasing error message strings.
Actually, although I have made clear that I don't like the change, if you want to merge it I'm fine, as you are right that tests are passing. Worst case, if the size goes up ever again, we shorten back the strings again.

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1137,405 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @param _amount Amount of tokens to be transferred from the msg.sender to the Court accounting
+    */
+    function _depositSenderAmount(ERC20 _token, uint256 _amount) internal {
+        if (_amount > uint256(0)) {
```
I think it's a good practice not to use constants as is. This shouldn't increase the tx costs, it is an obvious cast.

bingen replied:
But what's the benefit? A zero is a zero, right? In my opinion it just makes the code uglier.
Anyway, as good practices are quite subjective, I guess we should formalize them, compile them and agree to follow them. As far as I know we have never done that in Aragon (try, for instance, to `grep "> 0"` in Finance app...

facuspagnuolo replied:
No strong opinions seriously, but comparing two numbers without casting doesn't look good IMO, then if we agree on doing sth like `termId > uint64(0)`, we should follow that for `uint256` as well.

bingen replied:
Well, taking into account that 256 is the base for the EVM, I don't think it's the same doing it for `uint256` than for other sizes (I don't know what exactly the compiler does, I'll check), but I wouldn't do `uint64(0)` either, haha.
Worst case, again because `uint256` is native to the EVM, I would do `uint256(termId) > 0`, afraid of doing 2 casts, one down and another one up (again I have to check), but I don't like it either.
I don't see how comparing to zero can become an issue without a cast. Am I missing something? In that case, as I said, we should review all aragonOS and aragon-apps, because they are full of comparisons to zero without a cast.

facuspagnuolo replied:
True, it doesn’t make sense casting a zero to uint64 🤦‍♂
I’ll clean those up making sure we cast non-uint256 to uint256 when comparing numbers 👍

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +334,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         _ensureTermRandomness(draftTerm);
-        // as we already allow to move drafting to later terms, if current term has gone
-        // more than 256 blocks beyond the randomness BN, it will have to wait until next term
-        require(draftTerm.randomness != bytes32(0), ERROR_TERM_RANDOMNESS_UNAVAIL);
```
It is still needed and we are doing it in `_ensureTermRandomness ` (see https://github.com/aragon/aragon-court/pull/103/files#diff-4bb317b489994cf6f4c4b7ed11c18590R901)

bingen replied:
Thanks! I don't know how I missed that, sorry.

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +334,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-        }
+        // Draft the min number of jurors between the one requested by the sender and the one requested by the disputer
+        uint64 jurorsToBeDrafted = round.jurorsNumber - round.selectedJurors;
+        uint256 jurorsRequested = jurorsToBeDrafted < _maxJurorsToBeDrafted ? jurorsToBeDrafted : _maxJurorsToBeDrafted;
```
Given that it isn't a problem yet, don't you prefer code clarity than more space in the stack? IMO we were mutating the meaning of `jurorsRequested` before, and now we are modeling both things, the number of jurors pending to be drafted and the number requested by the user

izqui replied:
I agree with Facu that we shouldn't optimize for fewer stack variables unless absolutely necessary. Also these sort of optimizations could be done at a later point, once the test suite is frozen and can optimize all these things knowing that we won't break anything

bingen replied:
But again: the way I see it is that we are not optimizing, we are de-optimizing instead something that was already optimized. The supposed benefit is to "model `jurorsToBeDrafted`", but I don't see any modeling here, I just see a temporary variable that is only used in the line immediately below and never ever again in the +1K lines of code, and that may use a stack slot.
Let's fix [this](https://github.com/aragon/aragon-court/pull/103/files#r321671233) anyway to make sure that it is indeed unnecessary and if we can do it keeping this temporary variable I'll forget about it.

facuspagnuolo replied:
See 66194cc394a25683833d9e32e0d752cb4929e8ca

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+    event RulingAppealConfirmed(uint256 indexed disputeId, uint256 indexed roundId, uint64 indexed draftTermId, uint256 jurorsNumber);
     event RulingExecuted(uint256 indexed disputeId, uint8 indexed ruling);
-    event RoundSlashingSettled(uint256 indexed disputeId, uint256 indexed roundId, uint256 collectedTokens);
+    event PenaltiesSettled(uint256 indexed disputeId, uint256 indexed roundId, uint256 collectedTokens);
```
The thing is that penalties are settled for a group of users and rewards are settled individually

bingen replied:
Good point!

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+        uint64 coherentJurors;      // Number of drafted jurors that voted in favor of the dispute final ruling
+        uint64 settledJurors;       // Number of jurors whose rewards were already settled
+        uint256 collectedTokens;    // Total amount of tokens collected from losing jurors
+        bool settledPenalties;      // Whether or not penalties have been settled for a round
```
Fixed in 5dccdf737b624d78cfcfc45ca9cc73ff19204f68

bingen replied:
Hm, no, I don't think it's fixed: the four `uint64` just above will occupy one slot, so the `bool` will be alone again. Can you double-check, please?

facuspagnuolo replied:
I forgot the `triggeredBy` part 😅 
Fixed in 66194cc394a25683833d9e32e0d752cb4929e8ca

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +335,154 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-        public
-        ensureTerm
-    {
+    function draft(uint256 _disputeId, uint64 _maxJurorsToBeDrafted) external disputeExists(_disputeId) ensureTerm {
```
Optimized in 5dccdf737b624d78cfcfc45ca9cc73ff19204f68

bingen replied:
I still see `round.selectedJurors` repeated and `_round.jurorsNumber` inside `_draft` function.

facuspagnuolo replied:
Note that we have to re-load `round.selectedJurors` after calling `_draft` since it is being updated. Apart from that I tried passing both `jurorsNumber` and `selectedJurors` to `_draft` but we can only pass one of them due to a `stack too deep` issue, see 66194cc394a25683833d9e32e0d752cb4929e8ca

bingen replied:
> Note that we have to re-load round.selectedJurors after calling _draft since it is being updated. 

Indeed, having the call 1100 lines away doesn't help to spot that ;-)

I tried removing the temporary variable `jurorsToBeDrafted` and it didn't help either this time :man_shrugging:  (in a previous test I did when I started the conversation it did). Anyway, last 2 comments about it:

- I don't think its name is accurate. The jurors that are going to be drafted are the result of `requestedJurors`, so `jurorsToBeDrafted` sounds misleading to me (and too similar to the concept of requested jurors).
- In 4 lines we have: `jurorsNumber`, `selectedJurors`, `requestedJurors` and `_maxJurorsToBeDrafted`. Adding a 5th one `jurorsToBeDrafted` doesn't help in code clarity for me at all but I guess each mind works different.

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -17,204 +18,273 @@ import "@aragon/os/contracts/common/Uint256Helpers.sol";
+        uint64 revealTerms;         // revealing period duration in terms
+        uint64 appealTerms;         // appealing period duration in terms
+        uint64 appealConfirmTerms;  // confirmation appeal period duration in terms
+        uint16 penaltyPct;          // pre-mille that will be used to compute the tokens to be locked for a juror on a draft
```
Fixed in 66194cc394a25683833d9e32e0d752cb4929e8ca

bingen replied:
The typo? ("pre" vs "per") :grimacing:

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1139,415 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+
+        // If given term is before the appeal confirmation end term and the max number of appeal round hasn't been reached, round appeal can still be confirmed
+        uint64 appealConfirmationEndTerm = appealConfirmationStartTerm + config.appealConfirmTerms;
+        if (_termId < appealConfirmationEndTerm && _roundId < config.maxRegularAppealRounds) {
```
I like the idea, fixed in a81c2d16686db97fd7736a52d71709b83a1e19a6

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1136,405 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    */
+    function _getFinalRoundJurorsNumber(uint64 _termId) internal view returns (uint64) {
+        // The registry guarantees its total active balance will never be grater than
+        // `2^64 * minJurorsActiveBalance / FINAL_ROUND_WEIGHT_PRECISION`. Thus, the
```
It is not implemented yet, but I already created an issue to tackle it, see https://github.com/aragon/aragon-court/issues/102. I didn't want to mention here that we will ensure that cause we will probably forget to update it later, but since we do are going to tackle that issue before deploying the Court, I think we are safe to have this comment from now on

bingen replied:
For sure, I was not aware of that issue. Thanks for the clarification!

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -263,264 +333,153 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        // TODO: ensure round exists
+        // Ensure given round can be appealed. Note that if there was a final appeal the adjudication state will be 'Ended'.
         Dispute storage dispute = disputes[_disputeId];
+        _ensureAdjudicationState(dispute, _roundId, AdjudicationState.Appealing);
```
Both scenarios are contemplated within `_adjudicationStateAt`, see https://github.com/aragon/aragon-court/pull/103/commits/a81c2d16686db97fd7736a52d71709b83a1e19a6

bingen replied:
Ok, I didn't have that commit, but still, I'm a little bit lost. 
Before, `_checkAdjudicationState` would revert if dispute state was not Adjudicating.
Now, it seems its replacement `_ensureAdjudicationState` won't: it calls `_adjudicationStateAt` that returns ended in a lot of different scenarios that I'm still trying to understand.
`_ensureAdjudicationState` is called by `_ensureFinalRuling`, which in urn is called by `settlePenalties`.
It seems to me that before `settleRoundSlashing` (equivalent to `settlePenalties`) would revert if called for a dispute that is not Adjudicating (for instance Executed), but `settlePenalties` won't, and that this is a mistake.
TBH, I'm not sure at all, I will keep reviewing this logic, but I want to raise this up here to make sure we are not missing anything.
`_ensureAdjudicationState`

facuspagnuolo replied:
> Before, `_checkAdjudicationState` would revert if dispute state was not Adjudicating.
Now, it seems its replacement `_ensureAdjudicationState` won't: it calls `_adjudicationStateAt` that returns ended in a lot of different scenarios that I'm still trying to understand.

That doesn’t necessarily mean that it doesn't revert, it will depend on the state you're trying to ensure. However, I don't think I follow completely what you're trying to point out here

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1136,405 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @return Jurors number for final rounds for the given term
+    */
+    function _getFinalRoundJurorsNumber(uint64 _termId) internal view returns (uint64) {
+        // The registry guarantees its total active balance will never be grater than
```
Fixed in 3480bc3f0e17bc4411072d8bd511d6071c37201c

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1136,405 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        if (jurorsNumber % 2 == uint256(0)) {
+            jurorsNumber++;
+        }
+        return jurorsNumber;
```
Yea, I always suggest avoiding the not explicit returns unless it is completely obvious for example when returning multiple values. We can discuss this topic as well, the importance is to be consistent.

bingen replied:
Yes, I'm completely neutral about it, but I like the idea of trying to be consistent.

bingen commented on contracts/Court.sol:
```diff
@@ -263,264 +333,158 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @param _roundId Identification number of the dispute round being appealed
+    * @param _ruling Ruling appealing a dispute round in favor of
+    */
+    function appeal(uint256 _disputeId, uint256 _roundId, uint8 _ruling) external disputeExists(_disputeId) ensureTerm {
```
Don't we need to check that round exists too?

facuspagnuolo replied:
It is being checked by `_ensureAdjudicationState`

bingen commented on contracts/Court.sol:
```diff
@@ -263,264 +333,158 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @param _ruling Ruling being confirmed against a dispute round appeal
+    */
+    function confirmAppeal(uint256 _disputeId, uint256 _roundId, uint8 _ruling) external ensureTerm {
+        // TODO: ensure dispute exists
```
Same as above? Only the dispute? What about the round?

facuspagnuolo replied:
It is being checked by `_ensureAdjudicationState`

bingen commented on contracts/Court.sol:
```diff
@@ -778,351 +659,435 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        // Check if there was a final ruling already cached
         Dispute storage dispute = disputes[_disputeId];
-        DisputeState disputeState = dispute.state;
+        if (uint256(dispute.finalRuling) > 0) {
```
Damn it! I thought I convinced you to completely remove those castings. :stuck_out_tongue_winking_eye: 
I would like to understand what the danger is not to use them, and maybe we should raise it in Aragon core, as we have some like this one too, for instance these 2 ones:
https://github.com/aragon/aragon-apps/blob/master/apps/finance/contracts/Finance.sol#L272

facuspagnuolo replied:
I thought we agreed on making explicit castings unless it was obvious (e.g. `uint256(0)`). IMO it's a way to avoid mistakes and enforce us to think twice when doing operations with non-uint256 values. This scenario is trivial, but as we agreed before, we prefer being consistent

bingen replied:
To me it makes sense if there are 2 variables, but having a zero in one side it looks like extra noise.
Being consistent comprises Aragon core (OS and apps) code?

facuspagnuolo replied:
With "being consistent" I meant that if we do it for two variables, I'd do it even though one of those variables is a zero as well. On the other hand, IMO we should use the style guides in all our solidity codebase once we have them, but that's another topic.

bingen replied:
> I'd do it even though one of those variables is a zero as well

Well, 0 is a constant, not a variable :stuck_out_tongue_winking_eye: 

> we should use the style guides in all our solidity codebase once we have them

I completely agree, and as so far this is not done anywhere else in our codebase, I wouldn't do it here either, otherwise we'll start splitting styles.
But as I said off-line, proceed as you want.

bingen commented on contracts/Court.sol:
```diff
@@ -1180,15 +1145,421 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @param _roundId Identification number of the dispute round to be checked
+    * @param _state Expected adjudication state for the given dispute round
+    */
+    function _ensureAdjudicationState(Dispute storage _dispute, uint256 _roundId, AdjudicationState _state) internal view {
```
I'm no native English so this may not make any sense, but so far we were using `ensure` instead of `check` when we were making sure, so if the check was not holding true, we would change the state so it does.
I haven't reviewed all its internal calls yet, but as it's `view` I guess it's no "making sure" of anything.

facuspagnuolo replied:
Not "ensure" what you mean 😝

bingen replied:
Lol!! :laughing:

bingen replied:
More seriously, to clarify:
- check -> view, only check
- ensure -> non-view, checks and then modifies state if needed

facuspagnuolo replied:
Fixed in 87c7b3b460b05cfa22f0fb441c8e3c721a2ddafa

bingen commented on contracts/Court.sol:
```diff
@@ -1180,15 +1134,423 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    */
+    function _getConfigAt(uint64 _termId) internal view returns (CourtConfig storage) {
+        // If the given term is lower or equal to the last ensured Court term, it is safe to use a past Court config
+        uint64 lastEnsuredTermId = termId;
```
Why this?

facuspagnuolo replied:
Not sure what you're asking but this line intends to make sure we load the last ensured term only once

bingen replied:
My question is: why are we renaming `termId` to this temporary variable `lastEnsuredTermId`, instead of just using `termId`?

izqui replied:
Again I think this is mostly bikeshedding, but in this case I think specifying the name is a good idea since this function takes `_termId` as an argument and its meaning completely different from the `termId` state variable. Also `lastEnsuredTermId` is used more than once, so it's a good optimization as well.

bingen commented on contracts/Court.sol:
```diff
@@ -1180,15 +1134,423 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        }
+
+        // If the given term is in the future but there is a config change scheduled before it, use the incoming config
+        if (configChangeTermId <= _termId) {
```
Can't we just replace both conditions by:
```
    if (_termId > termId && _termId => configChangeTermId) {
        return _getConfigSafeAt(configChangeTermId);
    }
```

facuspagnuolo replied:
I don't think it will work since you will be skipping past configs, you will end up accessing either future scheduled configs or the last ensured one.

bingen replied:
Yes,, you are right! Sorry for the confusion.

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -679,97 +580,66 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+        emit AppealDepositSettled(_disputeId, _roundId);
+
+        // If the appeal wasn't confirmed, return the entire deposit to appeal maker
+        address appealMaker = appeal.maker;
```
See below, note that we are also using it [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.sol#diff-4bb317b489994cf6f4c4b7ed11c18590R611), [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.sol#diff-4bb317b489994cf6f4c4b7ed11c18590R621), and [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.sol#diff-4bb317b489994cf6f4c4b7ed11c18590R627)

bingen replied:
Yes, I see the 3 occurrences, but the thing is that they won't ever happen in the same call, so you will always load it once (unless I'm missing something).

facuspagnuolo replied:
True, good point. Fixed in https://github.com/aragon/aragon-court/pull/103/commits/12cb8eaf82b0ffa43e1bb262fcf37acfa800f24c

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -778,351 +648,435 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+    * @dev Tell the last ensured term identification number
+    * @return Identification number of the last ensured term
+    */
+    function getLastEnsuredTermId() external view returns (uint64) {
```
Not sure I follow the scenario you are describing. IMO this is what `termId` actually means, it's the last ensured term since it may be outdated.

bingen replied:
Ok, again, proceed as you want, it's just a name, but to make my point clear:  if `termId` means "the last ensured term", I don't see the need for adding `Ensured` as it would be just confusing, like there are 2 different concepts, `termId` and `ensuredTermId`.
But I don't think `termId` is "ensured" at all, as we don't know if all needed term transitions have taken place.
To me this is just and old plain getter:
```
    function getLastEnsuredTermId() external view returns (uint64) {
        return termId;
    }
```
And it's a common practice to name getters as `get` + variable name. So if you really want to call this with that (IMO) unnecessary long name instead of the classic `getTermId`, I would then rename the variable to `lastEnsuredTermId`.

[Suggestion for our style guide, section naming conventions: plain getters are called `get + variable name`]

izqui replied:
This is definitely a bit of bikeshedding at this point, but what about `getLastTransitionedTermId`?

bingen replied:
I agree it's bikeshedding. Actually I don't think we even need this function at all, as we could have kept `termId` public and use the automatically generated getter.
[Another thing to discuss for our style guide: do we use public state variables and auto-generated getters? Or do we declare them all internal and explicitly write getters?]

facuspagnuolo replied:
I like `getLastTransitionedTermId` and also agree about the getters pattern, I think the whole problem comes as a consequence of the state variable being called simply `termId` which is not always up-to-date. What about renaming the variable to `lastTransitionedTermId` as well?

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -778,351 +648,435 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
-            _term.randomness = _getTermRandomness(_term);
+        // Update the number of round settled jurors. Note that we don't need to use SafeMath here since the highest number of jurors to be
+        // settled for a round could be the `jurorsNumber` itself, which is a uint64 value.
+        _round.settledJurors = uint64(roundSettledJurors + batchSettledJurors);
```
Yes but as I mentioned in the PR description:

> Update `settledJurors` even for the last batch to ensure consistency

I think it's important to ensure completeness here, I could be useful in the future if we need to debug sth

izqui commented on test/court/court-disputes.js:
```diff
@@ -0,0 +1,265 @@
+          const jurorRewards = (draftFee.plus(settleFee)).mul(jurorsNumber)
+          const requiredCollateral = jurorFees.plus(heartbeatFee).plus(jurorRewards)
+
+          beforeEach('deposit collateral', async () => {
```
```suggestion
          beforeEach('approve fee amount', async () => {
```

izqui commented on test/court/court-disputes.js:
```diff
@@ -0,0 +1,265 @@
+            assert.equal(collectedTokens.toString(), 0, 'round collected tokens should be zero')
+          })
+
+          it('transfers the collateral to the court', async () => {
```
```suggestion
          it('transfers fees to the court', async () => {
```

izqui commented on test/court/court-disputes.js:
```diff
@@ -0,0 +1,265 @@
+          })
+        })
+
+        context('when the creator does not deposit enough collateral', () => {
```
```suggestion
        context('when the creator doesn\'t have enough fee tokens approved', () => {
```

izqui commented on test/court/court-disputes.js:
```diff
@@ -0,0 +1,265 @@
+            assert.equal(previousSenderBalance.minus(expectedDisputeDeposit).toString(), currentSenderBalance.toString(), 'sender balances do not match')
+          })
+
+          it(`transitions ${expectedTermTransitions} terms`, async () => {
```
We should check that whoever calls heartbeat for the term the dispute can be drafted is assigned the `heartbeatFee`

facuspagnuolo replied:
Added in https://github.com/aragon/aragon-court/pull/103/commits/e470083f72b9ebcd40d808059ada10b68782ed63#diff-09541abf0e82ec4183242cf0317e7597R350-R371

izqui commented on test/court/court-draft.js:
```diff
@@ -0,0 +1,438 @@
+        disputeId = await courtHelper.dispute({ jurorsNumber, draftTermId, disputer })
+      })
+
+      const itDraftsRequestedRoundInOneBatch = (term, jurorsToBeDrafted) => {
```
We should test that the drafter is assigned the `draftFee` for each juror drafted

facuspagnuolo replied:
I think we are doing it [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.sol#diff-09541abf0e82ec4183242cf0317e7597R118-R136), did I miss something?

izqui commented on test/court/court-voting.js:
```diff
@@ -0,0 +1,387 @@
+            const expectedWinningOutcome = highOutcomeTally > lowOutcomeTally ? OUTCOMES.HIGH : OUTCOMES.LOW
+            assert.equal(winningOutcome.toString(), expectedWinningOutcome, 'winning outcome does not match')
+          })
+          })
```
```suggestion
        })
```

izqui commented on test/court/court-appeal.js:
```diff
@@ -0,0 +1,363 @@
+
+                context('when the appeal maker has enough balance', () => {
+                  beforeEach('mint fee tokens for appeal maker', async () => {
+                    await courtHelper.feeToken.generateTokens(appealMaker, bigExp(1e6, 18))
```
Can we get these amounts from a getter in the Court?

facuspagnuolo replied:
Added in https://github.com/aragon/aragon-court/pull/103/commits/14da6572805f5ba6187ca4e832e38af7595ed58b

izqui commented on test/court/court-appeal.js:
```diff
@@ -0,0 +1,363 @@
+      })
+
+      // TODO: this scenario is not implemented in the contracts yet
+      context.skip('when the given round is not valid', () => {
```
What error is the contract giving for this now?

facuspagnuolo replied:
Actually I forgot to remove this skips, they were already fixed :)

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/cce6087153d97426b4b1c641862268096dea3910

izqui commented on test/court/court-confirm-appeal.js:
```diff
@@ -0,0 +1,361 @@
+
+                  context('when the appeal taker has enough balance', () => {
+                    beforeEach('mint fee tokens for appeal taker', async () => {
+                      await courtHelper.feeToken.generateTokens(appealTaker, bigExp(1e6, 18))
```
Can we get these amounts from a getter in the Court as well?

facuspagnuolo replied:
Added in https://github.com/aragon/aragon-court/pull/103/commits/14da6572805f5ba6187ca4e832e38af7595ed58b

izqui commented on test/court/court-settle.js:
```diff
@@ -0,0 +1,790 @@
+
+        const itFailsToSettleRewards = (roundId) => {
+          it('fails to settle rewards', async () => {
+            await assertRevert(court.settleReward(disputeId, roundId, juror1000), 'CT_ROUND_PENALTIES_NOT_SETTLED')
```
Nit: should we make `juror1000` an argument to `itFailsToSettleRewards`?

facuspagnuolo replied:
Actually here it could be any juror, will declare a proper account variable for this

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/cce6087153d97426b4b1c641862268096dea3910

izqui commented on test/court/court-settle.js:
```diff
@@ -0,0 +1,790 @@
+              assertEvent(receipt, 'RulingExecuted', { disputeId, ruling: expectedFinalRuling })
+            })
+
+            it('executes the associated airbitrable', async () => {
```
even though 'airbitrable' sounds amazing... 😉

```suggestion
            it('executes the associated arbitrable', async () => {
```

izqui commented on test/court/court-settle.js:
```diff
@@ -0,0 +1,790 @@
+                itSettlesPenaltiesProperly()
+
+              } else {
+                it ('reverts', async () => {
```
```suggestion
                it('reverts', async () => {
```

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1180,15 +1134,423 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
+
+        // If given term is before the appeal confirmation start term and the last round was not appealed yet, then the last round can still be appealed
+        Appeal storage appeal = round.appeal;
+        bool isLastRoundAppealed = !_existsAppeal(round.appeal);
```
Nice catch!

facuspagnuolo commented on test/helpers/court.js:
```diff
@@ -49,6 +144,194 @@ module.exports = (web3, artifacts) => {
+
+    async confirmAppeal({ disputeId, roundId, appealTaker = undefined, ruling = undefined }) {
+      // mint fee tokens for the appeal taker, if no taker was given pick the fifth account
+      if (!appealTaker) appealTaker = web3.eth.accounts[4 + roundId * 2]
```
The test script is running a ganache with 200 accounts. However, this should be fixed to access the 5th account. Nice catch!

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/cce6087153d97426b4b1c641862268096dea3910

facuspagnuolo commented on test/helpers/court.js:
```diff
@@ -49,6 +144,194 @@ module.exports = (web3, artifacts) => {
+      // call heartbeat function for X terms
+      await this.court.heartbeat(terms)
+      // advance 2 blocks to ensure we can compute term randomness
+      const { advanceBlocks } = require('../helpers/blocks')(web3)
```
Don't think so, will move it to the top :)

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/cce6087153d97426b4b1c641862268096dea3910

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -1141,9 +1095,9 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         // We make sure that when applying penalty pct to juror min stake it doesn't result in zero
         uint256 minJurorsActiveBalance = jurorsRegistry.minJurorsActiveBalance();
-        require(uint256(_penaltyPct) * minJurorsActiveBalance >= PCT_BASE, ERROR_WRONG_PENALTY_PCT);
+        require(minJurorsActiveBalance.pct(_penaltyPct) > 0, ERROR_INVALID_PENALTY_PCT);
```
Agree completely, let's fix it

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/37ca5d6cc8a7bcf44698ea860cd092b8ea1c9717

facuspagnuolo commented on contracts/Court.sol:
```diff
@@ -382,25 +377,29 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
         CourtConfig storage config = _getConfigAtDraftTerm(round);
-        _draft(jurorsRequested, _disputeId, round, draftTerm, config);
-        accounting.assign(config.feeToken, msg.sender, config.draftFee * jurorsRequested);
+        _draft(_disputeId, round, jurorsNumber, requestedJurors, draftTerm, config);
```
Nice catch, let's do that

facuspagnuolo replied:
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/37ca5d6cc8a7bcf44698ea860cd092b8ea1c9717

facuspagnuolo replied:
Actually had to rollback due to a `stack-too-deep` issue, will add a TODO to see this later

facuspagnuolo commented on test/court/court-initialize.js:
```diff
@@ -0,0 +1,29 @@
+      await assertRevert(courtHelper.deploy({ mockedTimestamp: NOW, firstTermStartTime: NOW - 1, termDuration: ONE_DAY }), 'CT_BAD_FIRST_TERM_START_TIME')
+    })
+
+    it('cannot use a penalty pct lower than 1% (1/10,000) ', async () => {
```
Fixed in https://github.com/aragon/aragon-court/pull/103/commits/f616da3eb5d3fef4b6910f2cfc29a8b4fd126c6e

facuspagnuolo commented on test/court/court-draft.js:
```diff
@@ -0,0 +1,438 @@
+          })
+        })
+
+        context('when the current term is after the randomness block number by more than 256 blocks', () => {
```
> We should also test that if we do a heartbeat we can indeed draft the round.

Added in (`e470083`)[https://github.com/aragon/aragon-court/pull/103/commits/e470083f72b9ebcd40d808059ada10b68782ed63#diff-09541abf0e82ec4183242cf0317e7597R353-R357]

> It would also be a neat test to draft half the jurors, shuffle the tree a little bit, advance a term and finish the draft.

We are doing sth pretty similar [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.js&hide-deleted-files=true#diff-09541abf0e82ec4183242cf0317e7597R139) when testing multiple batches. Note that we are always [transitioning terms](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.js&hide-deleted-files=true#diff-09541abf0e82ec4183242cf0317e7597R145) between each batch (see [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.js&hide-deleted-files=true#diff-c1dbeba95466d900f8fac13c1f3472f3R165-R172)). We are not shuffling the tree tho.

facuspagnuolo commented on test/helpers/court.js:
```diff
@@ -34,6 +56,79 @@ module.exports = (web3, artifacts) => {
+      return nextRoundJurorsNumber
+    }
+
+    getDraftDepositFor(jurorsNumber) {
```
Added in https://github.com/aragon/aragon-court/pull/103/commits/234267f73369608b19f2fe6dffddf898fb2b7a9b

facuspagnuolo commented:
@izqui I already addressed all your comments. Please take a look at the related changes, I'd recommend reviewing them by commit. I improved and added some other appeal tests in https://github.com/aragon/aragon-court/pull/103/commits/e2cac0b66b878367f550bdcec00f1019459115bd, and also noted that we may be missing the final round reduction [here](https://github.com/aragon/aragon-court/pull/103/files?file-filters%5B%5D=.sol#diff-4bb317b489994cf6f4c4b7ed11c18590R1251-R1252). LMKYT

izqui commented on contracts/Court.sol:
```diff
@@ -1235,6 +1235,7 @@ contract Court is IJurorsRegistryOwner, ICRVotingOwner, ISubscriptionsOwner, Tim
 
         // Otherwise, return the times the active balance of the juror fits in the min active balance, multiplying
         // it by a round factor to ensure a better precision rounding.
+        // TODO: review, we are not using the final round discount here
```
Good catch. I don't think we should have it here though, but it should definitely be in `settleReward`  when assigning fees for coherent jurors in the final round https://github.com/aragon/aragon-court/blob/e2cac0b66b878367f550bdcec00f1019459115bd/contracts/Court.sol#L591

And we should definitely have a test for it. Maybe we should work on it on a specific test file for accounting?

bingen replied:
Hm, I don't get it. `round.jurorFees` already has the discount applied, right?

facuspagnuolo replied:
True, I was confused. The problem was that we are not testing it 😂
